### PR TITLE
Add docs on updating PRs using git

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This PR adds documentation on how to update PRs using git by amending or squashing.

This is in the context of #8503.

### CODE REVIEW

File: doc/devel/gitwash/development_workflow.rst [hunk 1]
```diff
@@ -172,6 +172,71 @@ When you are ready to ask for someone to review your code and consider a merge:
    pull request message.  This is still a good way of getting some preliminary
    code review.
 
+Update your pull request
+========================
+
+Often, you'll want to update your contribution based on the feedback in the
+pull request. While it's possible to simply push additional commits to your
+feature branch, this will clutter the history. Not every step in the
+development of a feature is worth to be recorded. In contrast, a clean
```
Comment (phobson) [inline doc/devel/gitwash/development_workflow.rst]: perhaps "not ever step...is worth recording"

File: doc/devel/gitwash/development_workflow.rst [hunk 2]
```diff
@@ -172,6 +172,71 @@ When you are ready to ask for someone to review your code and consider a merge:
...
+history with one or a few thematically separted commits helps to keep it
```
Comment (phobson) [inline doc/devel/gitwash/development_workflow.rst]: separated

File: doc/devel/gitwash/development_workflow.rst [hunk 3]
```diff
@@ -172,6 +172,71 @@ When you are ready to ask for someone to review your code and consider a merge:
...
+Ammend an existing commit
+   Use this if you have don't need to keep track of the incremental changes
+   to your feature.
+Squash multiple commits
+   Use this to cleanup the history when you've already made multiple commits
+   to your feature branch.
+
+Both techniques rewrite the history of your feature branch. While generally
+one has to be careful with rewriting the history, this is intended and
+perfectly fine for updates to your feature branch.
```
Comment (phobson) [inline doc/devel/gitwash/development_workflow.rst]: perhaps: "perfectly fine for commits that are currently on your feature branch only".

a point that should be clear is that commits already in master shouldn't be touched.

**Comment by ImportanceOfBeingErnest** [comment]
:+1: This exactly what I need. Just a question: Often people review certain lines of code. How are those treated? Will the review then point to a nonexisting line after ammending?

File: doc/devel/gitwash/development_workflow.rst [hunk 4]
```diff
@@ -172,6 +172,71 @@ When you are ready to ask for someone to review your code and consider a merge:
...
+updating pull requests.
+
+Ammending the last commit
+-------------------------
+You have pushed a single commit for your feature and now want to update it.
+
+On your feature branch::
+
+   # make your updates
+   git commit --all --amend --no-edit
```
Comment (phobson) [inline doc/devel/gitwash/development_workflow.rst]: Non-blocker: I'm a little nervous about recommending `--all`. I'd prefer to instruction people unfamiliar with git to add each of the files rather that possibly adding a bunch of unrelated files.

File: doc/devel/gitwash/development_workflow.rst [hunk 5]
```diff
@@ -172,6 +172,71 @@ When you are ready to ask for someone to review your code and consider a merge:
...
+Squashing the last N commits
+----------------------------
+You have N commits in your feature branch (e.g. N=3) and want to turn them
+into a single commit.
+
+On your feature branch::
+
+    git reset --soft HEAD~3
+    git commit
+    git push --force
```
Comment (phobson) [inline doc/devel/gitwash/development_workflow.rst]: does this achieve the same thing as: 

```
git checkout feature-branch
git rebase -i HEAD~8 # opens text editor, check your gitconfig
# use the fixup option on all but the first line
# save and exit text editor (for vim that's ":wq"
git push origin feature-branch --force
```

**Comment by phobson** [comment]
@ImportanceOfBeingErnest when an individually reviewed line changes, the GitHub UI collapses the comment. You can click on a button labeled "show outdated comment" to see the history. Your PR from the pie docs has this: https://github.com/matplotlib/matplotlib/pull/10953#issuecomment-379038829

**Comment by timhoffm** [comment]
You can see the outdated comment and the few lines of code segment. However, naturally the "View Changes" button leads nowhere as there is simply no commit associated anymore. This is inherent. We wanted to throw away the preliminary commits.

File: doc/devel/gitwash/development_workflow.rst (hunk 4 above)
Comment (jklymak) [reply doc/devel/gitwash/development_workflow.rst]: Yes definitely agree.  For better or worse I have a bunch of test files in my local dev repository and if I used -all I’d be in a lot of trouble.

File: doc/devel/gitwash/development_workflow.rst (hunk 5 above)
Comment (jklymak) [reply doc/devel/gitwash/development_workflow.rst]: Yeah it does.  But I think it requires one fewer editor opens.    I somewhat prefer the rebase option.

Comment (timhoffm) [reply doc/devel/gitwash/development_workflow.rst]: I think so, but you're never quite sure with git. I took this from https://stackoverflow.com/a/5201642/2726279

**Comment by ImportanceOfBeingErnest** [comment]
Ok, so this is in anyway desireable to do. 

There is a section down below at that page called "Rewriting commit history", should that not somehow unified with the new changes?

<hr>

A further note. I think the problems I had with previous PRs trying to squash was that there were other commits from other people in between those I wanted to merge. 

So when I did 

    git rebase -i HEAD~4

the commits shown in the editor were not just my last commits from my branch, but simply the last 4 commits to the matplotlib. E.g.

```
pick xxxxxxxxx myupdate 1
pick yyyyyyyyy someone else's update
pick xyyxyxyxy myupdate 2
pick zzzzzzzzz another person's update
```

So it wasn't clear what to do then. 

The explanation in the "Rewriting commit history" has the drawback that it isn't clear which of the mentionned commits would be mine. Also there is the word "last", for which I wasn't sure if it is meant temporarily (newest) or spacially (bottom-most?). 

Also `git log --oneline` continuously lets my git become unresponsive. I did look the website's numbers to find out instead. Maybe that is a known issue? But it makes it even harder to determine which commit to specify in 

    git rebase -i yxzxyzxy

<hr>

Concerning ammend, is this still an option if you already have several commits (due to forgetting to ammend previously or due to not knowing about the existance of this option)?

**Comment by jklymak** [comment]
@ImportanceOfBeingEarnest that’s almost certainly because you did a merge commit before you attempted rebasing.  I’ve done that by accident a few times and if that happens you better hope you have a backup.  That eventuality would ideally be covered here as well. 

I agree that this should be homogenized with the rewriting history. 

Btw I’m not sure we get to edit any of this. I think it gets pulled from the gitwash repository.

File: doc/devel/gitwash/development_workflow.rst (hunk 5 above)
Comment (timhoffm) [reply doc/devel/gitwash/development_workflow.rst]: In my experience rebasing is more difficult to understand than soft reset. That's why I chose this approach. I still feel like I cannot explain rebasing reasonably simple.

**Comment by timhoffm** [comment]
If you've accidentially merged, you can still reset --hard to your branch before the merge commit.

The "Rewriting commit history" section is also part of development_workflow.rst. It's editable in principle, but I don't feel comfortable with the topic.

@ImportanceOfBeingErnest Concerning amend when you already have multiple commits. You an change single ones but the number of commits will stay. How to amend the last one is explained above. To amend an earlier commit you have to edit and commit-amend as part of a (interactive) rebase.



@ImportanceOfBeingErnest Concerning amend when you already have multiple commits. You an change single ones but the number of commits will stay. How to amend the last one is explained above. To amend an earlier commit you have to edit and commit-amend as part of a (interactive) rebase.

**Comment by jklymak** [comment]
That `reset --soft` is really useful.  Definitely get that in.  

But again, do we actually get to edit this, or is it just copied from https://github.com/matthew-brett/gitwash/tree/master/gitwash ?  I think the changes need to be proposed there, but I'm fuzzy on whether that is really the case.

**Comment by tacaswell** [comment]
This should be opened up against upstream gitwash or your changes will at somepoint be blown away and you will be sad.

**Comment by jklymak** [comment]
@timhoffm I'm closing this as this is the wrong repository for this.  But I think the actual work is great!
